mlir/lib/Transforms/lab4/sharapov_georgiy/sCeilDiv.cpp

```diff
@@ -0,0 +1,72 @@
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Tools/Plugins/PassPlugin.h"
+
+using namespace mlir;
+
+namespace {
+class sCeilDivPass : public PassWrapper<sCeilDivPass, OperationPass<ModuleOp>> {
+public:
+  StringRef getArgument() const final { return "sharapov_ceildiv"; }
+  StringRef getDescription() const final {
+    return "breaks arith.ceildivui and arith.ceildivsi operations into arith. "
+           "calculations";
+  }
+
+  void runOnOperation() override {
+    getOperation()->walk([&](Operation *op) {
+      if (auto ceilDivUI = dyn_cast<arith::CeilDivUIOp>(op)) {
+        replaceCeilDivUI(ceilDivUI);
+      } else if (auto ceilDivSI = dyn_cast<arith::CeilDivSIOp>(op)) {
+        replaceCeilDivSI(ceilDivSI);
+      }
+    });
+  }
+
+private:
+  void replaceCeilDivUI(arith::CeilDivUIOp op) {
+    OpBuilder builder(op);
+    Location loc = op.getLoc();
+    Value a = op.getLhs();
+    Value b = op.getRhs();
+
+    Value one =
+        builder.create<arith::ConstantIntOp>(loc, 1, builder.getI32Type());
+    Value add = builder.create<arith::AddIOp>(loc, a, b);
+    Value sub = builder.create<arith::SubIOp>(loc, add, one);
+    Value div = builder.create<arith::DivUIOp>(loc, sub, b);
```

Comment:
The difference between methods is only in one instruction. Would be nice to write one method with selections depending on input type to reduce code duplication.